### ISSUE TITLE
piper fails to run

### ISSUE THREAD

**Issue by bwanaaa** [libratbag/piper#484]
Installed libratbag and piper as per instructions
Mouse is Logitech G703 and Logitech power charging mousepad

**Information**

- `ratbagd` version= 0.13
- Piper version: 
`piper --version`
 gives:
```
Traceback (most recent call last):
  File "/usr/bin/piper", line 3, in <module>
    import gi
ModuleNotFoundError: No module named 'gi'

```
`sudo apt-get install python3-gi`
gives
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
python3-gi is already the newest version (3.26.1-2ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 75 not upgraded.
```

- Distribution: Ubuntu 18.04 LTS
- Kernel version Linux 4.15.0-52-generic x86_64 GNU/Linux

**Describe the bug**```

typing 'piper in the command line gives
```
Traceback (most recent call last):
  File "/usr/bin/piper", line 3, in <module>
    import gi
ModuleNotFoundError: No module named 'gi'
```

even though I do have gi installed

**Logs**
Please start the daemon with verbosity and reproduce the issue.
?
```
$ ratbagd --verbose=raw
Initializing libratbag
ratbag debug: New device: Logitech USB Receiver
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 046d:c53a
ratbag debug: New device: Logitech USB Receiver
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 046d:c53a
ratbag debug: New device: Logitech USB Receiver
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 046d:c53a
ratbag debug: New device: ASUS ASUS ProArt Calibration Device
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 0b05:187f
ratbag debug: New device: Razer Razer Tartarus V2
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:022b
ratbag debug: New device: Razer Razer Tartarus V2
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:022b
ratbag debug: New device: Razer Razer Tartarus V2
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:022b
ratbag debug: New device: Razer Razer BlackWidow Chroma
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:0203
ratbag debug: New device: Razer Razer BlackWidow Chroma
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:0203
ratbag debug: New device: Razer Razer BlackWidow Chroma
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:0203
DBus server ready

```

**Additional context**
Just to make sure piper was installed I tried installing again and got

```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
piper is already the newest version (0.3+git396-74e6c5a~ubuntu18.04.1).
0 upgraded, 0 newly installed, 0 to remove and 75 not upgraded.
```
**

**Comment by stephanlachnit** [comment]
Are you sure you are defaulting to your system installation of python? Some badly programmed software like Anaconda replaces your system installation. If your system isn't broken, you should be able to run `import gi` in a python console (open one with `$ python3`).

**Comment by FFY00** [comment]
It is worth to check the output of `which python3`.

**Comment by bwanaaa** [comment]
Thank you for your replies.
Here are some more terminal outputs that may help you decipher things.
After I installed anaconda 3, every time I open terminal I get (base) before my username. I think that means it's using the base environment. In that environment

```
(base) stefan@4930:~/Dropbox/aPython$ python -V
Python 3.7.6

(base) stefan@4930:~/Dropbox/aPython$ python
Python 3.7.6 (default, Jan  8 2020, 19:59:22) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import gi
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'gi'

```
even though gtk is there
```

(base) stefan@4930:~/Dropbox/aPython$ sudo apt-get install python3-gi
[sudo] password for stefan: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
python3-gi is already the newest version (3.26.1-2ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 75 not upgraded.

```

ok. so it seems i am using the python in the anaconda distro. but it is python3 so who cares if I am not using the system python? I'm using python3.

Anyway if I completely get back to the base python (conda deactivate) and try piper like so:
`stefan@4930:~/Dropbox/aPython$ piper`
I get a window with the cheese in the mouse trap that says it cannot find any devices

and btw, 
```
stefan@4930:~/Dropbox/aPython$ python -V
Python 2.7.17
```
So in the completely 'native' state where I am using the system python(2.717) and installed  piper, I get a different problem.

**Comment by stephanlachnit** [comment]
> ok. so it seems i am using the python in the anaconda distro. but it is python3 so who cares if I am not using the system python? I'm using python3.

Yeah Anaconda is the problem. I'm sorry I can't help here, but Anaconda is broken by design.
I can explain you why it doesn't work, but really just uninstall that crap pile of software that shouldn't exist on Linux (sorry-not-sorry for the rant). It's madness that they even provide a Linux version without at least one warning on their website.
Modern programs for Linux won't ship their dependencies - instead, distributions like Ubuntu have to make sure that the required dependencies are installed. This way, if program A and program B both require program C, it will only be installed _once_. However, Windows doesn't work like this, basically every program ships the libraries it needs (because they are just no standards for the locations), leading to multiple versions of libraries scattered around your system, and some programs don't even care about updating them, which btw creates tons of security issues. One can clearly see that Anaconda is designed for Windows, since it ships with all the Python modules it _thinks_ that are "necessary". Works fine on Windows, and I wouldn't really mind too much on Linux, **if** they would use a different binary like `conda_python3`. However, they don't, instead they _replace_ your system installation of Python (or rather, the symlink). You may ask yourself why this is a problem, but this means that your Python installation can only access the modules installed by Anaconda. And since Anaconda is only focused for Windows, tons of modules required for normal system operation are missing, including the gi module apparently. If your life is boring you can start to install all the packages via Anaconda's pip, but this will end in madness, since pip doesn't have a proper update system (it's just not designed for system packages).
tl;dr uninstall Anaconda, install the stuff you need from Anaconda from the Ubuntu repository, and if that stuff is too old consider upgrading to a more modern distro.
Don't get me wrong, you might not know this stuff so don't take it personal, but it just sucks that they tell you how great their software is when in fact it is simply system-breaking.

**Comment by bwanaaa** [comment]
Thank you for your replies. Here is the data you requested

```
stefan@4930:~/Dropbox/aPython$ lsusb
Bus 002 Device 004: ID 1532:0203 Razer USA, Ltd 
Bus 002 Device 003: ID 1532:022b Razer USA, Ltd 
Bus 002 Device 002: ID 8087:0024 Intel Corp. Integrated Rate Matching Hub
Bus 002 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 007 Device 004: ID 0b05:187f ASUSTek Computer, Inc. 
Bus 007 Device 003: ID 05e3:0610 Genesys Logic, Inc. 4-port hub
Bus 007 Device 002: ID 0bda:5420 Realtek Semiconductor Corp. 
Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 006 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 004 Device 002: ID 174c:3074 ASMedia Technology Inc. ASM1074 SuperSpeed hub
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 003: ID 046d:c53a Logitech, Inc. 
Bus 003 Device 002: ID 174c:2074 ASMedia Technology Inc. ASM1074 High-Speed hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 003: ID 0b05:17cf ASUSTek Computer, Inc. 
Bus 001 Device 002: ID 8087:0024 Intel Corp. Integrated Rate Matching Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

```
Also
```
stefan@4930:~/Dropbox/aPython$ sudo systemctl daemon-reload --verbose
systemctl: unrecognized option '--verbose'
stefan@4930:~/Dropbox/aPython$ sudo systemctl reload dbus.service --verbose
systemctl: unrecognized option '--verbose'
stefan@4930:~/Dropbox/aPython$ 

```


How do I restart the daemon with verbosity output?

**Comment by bwanaaa** [comment]
BTW, I COMPLETELY uninstalled anaconda, including the hidden .directories as well as all mention of anaconda from bashrc. Here is my PATH

```
echo $PATH
/usr/local/texlive/2019/bin/x86_64-linux:/home/stefan/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/SageMath/:/usr/local/cuda-10.1/bin

```
piper still gives me at mousetrap with cheese in it.

**Comment by bwanaaa** [comment]
Also in /usr/share/libratbag there is logitech-g703.device. In that file is:

```
[Device]
Name=Logitech G703
DeviceMatch=usb:046d:c087;usb:046d:4070
Driver=hidpp20
```

**Comment by bwanaaa** [comment]
and just to be sure

```
stefan@4930:~/Dropbox/aPython$ which python3
/usr/bin/python3
stefan@4930:~/Dropbox/aPython$ python3 -V
Python 3.6.9

```

**Comment by bwanaaa** [comment]
And here is the daemon with verbosity

```
sudo ratbagd --verbose=raw
Initializing libratbag
ratbag debug: New device: Logitech USB Receiver
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: device assigned driver hidpp20
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw6'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw7'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: - HID report ID 02
ratbag debug: - HID report ID 03
ratbag debug: - HID report ID 04
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw8'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: - HID report ID 10
ratbag debug: - HID report ID 11
ratbag debug: hidpp: device supports short reports
ratbag debug: hidpp: device supports long reports
ratbag raw: hidpp write:  11 ff 00 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag error:     USB error: Connection timed out (110)
write: Success
ratbag debug: New device: Logitech USB Receiver
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: device assigned driver hidpp20
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw6'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw7'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: - HID report ID 02
ratbag debug: - HID report ID 03
ratbag debug: - HID report ID 04
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw8'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: - HID report ID 10
ratbag debug: - HID report ID 11
ratbag debug: hidpp: device supports short reports
ratbag debug: hidpp: device supports long reports
ratbag raw: hidpp write:  11 ff 00 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:   10 ff 8f 00 18 01 00
ratbag debug:     HID++ error from the device (255): ERR_UNKNOWN (01)
ratbag debug: New device: Logitech USB Receiver
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: device assigned driver hidpp20
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw6'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw7'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: - HID report ID 02
ratbag debug: - HID report ID 03
ratbag debug: - HID report ID 04
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw8'.
ratbag debug: Parsing HID report descriptor
ratbag debug: Parsing HID report descriptor
ratbag debug: - HID report ID 10
ratbag debug: - HID report ID 11
ratbag debug: hidpp: device supports short reports
ratbag debug: hidpp: device supports long reports
ratbag raw: hidpp write:  11 ff 00 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag error:     USB error: Connection timed out (110)
write: Success
ratbag debug: New device: ASUS ASUS ProArt Calibration Device
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 0b05:187f
ratbag debug: New device: Razer Razer Tartarus V2
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:022b
ratbag debug: New device: Razer Razer Tartarus V2
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:022b
ratbag debug: New device: Razer Razer Tartarus V2
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:022b
ratbag debug: New device: Razer Razer BlackWidow Chroma
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:0203
ratbag debug: New device: Razer Razer BlackWidow Chroma
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:0203
ratbag debug: New device: Razer Razer BlackWidow Chroma
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: No data file found for 1532:0203
DBus server ready
```

**Comment by bwanaaa** [comment]
it sees the receiver
```
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc53a
ratbag debug: Logitech USB Receiver is device '/dev/hidraw7'.
```

but what are the errors about?

```
ratbag raw: hidpp write:  11 ff 00 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag error:     USB error: Connection timed out (110)
write: Success
```

**Comment by stephanlachnit** [comment]
The G-Powerplay not appearing in `lsusb` is a kernel issue afaik. Anyway, I don't really get why it isn't working, the G703 is not appearing in `lsusb`, `046d:c53a` is the USB-ID of the reveiver afaik.
This all points to a kernel issue, but I can't really help here, that's something @FFY00 and @bentiss know more about.

**Comment by FFY00** [comment]
What do you mean? The Powerplay shows up.
```
Bus 003 Device 003: ID 046d:c53a Logitech, Inc. 
```

@bwanaaa you need to upgrade to at least Linux 5.0 for the current device files to work. There were some changes in the `hid-logitech-dj` kernel driver in Linux 5.0 which we rely on.

If you really can't do that try this .device file:
```
[Device]
Name=Logitech G703
DeviceMatch=usb:046d:c53a
Driver=hidpp20

[Driver/hidpp20]
DeviceIndex=1
```
Please note that this is unsupported.

**Comment by bwanaaa** [comment]
YOWZA! I did thAt and you know...........................you;re RIGHT..................it works...AAARGH!!

**Comment by FFY00** [comment]
Great, but you might want to actually upgrade your kernel as that .device file \*is not\* supported.

**Comment by sudo-nano** [comment]
@stephanlachnit Thanks for your explanation about conda breaking piper. It saved me just now. I couldn't figure out why piper had suddenly stopped working. Lo and behold, the only variable was that I had installed conda since the last time it worked.
